tools/upload-json-dashboard.sh

```diff
@@ -0,0 +1,281 @@
+            ;;
+        '?' )
+            printf "$ERRFMT" $LINENO "invalid option: -$OPTARG" 1>&2
+            exit 1
+            ;;
+    esac
+done
+shift $((OPTIND -1))
+
+# ================================================================
+# Check required options.
+# ================================================================
+if [ -n "$DASH_JSON" ] && [ -z "$DASH_NAME" ] ; then
+    if jq --version 1>/dev/null 2>/dev/null ; then
+        # If jq is available, extract the variable name.
+        DASH_NAME=$(jq '.__inputs[0].name' "$DASH_JSON" | awk -F'"' '{print $2}')
+    fi
+fi
+# need bash 4 or later
+BASH_MAJOR=$(echo "$BASH_VERSION" | awk -F. '{print $1}')
+if (( BASH_MAJOR < 4 )) ; then
+    # Need 4 or later to support "declare -A".
+    printf "$ERRFMT" $LINENO "this version of bash is too old, must 4 or later: $BASH_VERSION" 1>&2
+    exit 1
+fi
+
+ERRCNT=0
+declare -A ROPTS=( ['-d']="$DASH_DS" ['-j']="$DASH_JSON" ['-g']="$DASH_URL" ['-n']="$DASH_NAME")
+for RKEY in "${!ROPTS[@]}" ; do
+    RVAL=${ROPTS[$RKEY]}
+    if [ -z "$RVAL" ] ; then
+        printf "$ERRFMT" \
+               $LINENO \
+               "missing required option: '$RKEY', see the help (-h) for more information" \
+               1>&2
+        ERRCNT=$(( ERRCNT += 1))
+    fi
+done
+if (( ERRCNT )) ; then exit 1 ; fi
+
+# ================================================================
+# Report the setup.
+# ================================================================
+if (( VERBOSE )) ; then
+    cat <<EOF
+Setup
+    BASH_VERSION    : $BASH_VERSION
+    DASH_AUTH       : $DASH_AUTH
+    DASH_DS         : $DASH_DS
+    DASH_FOLDER     : $DASH_FOLDER
+    DASH_JSON       : $DASH_JSON
+    DASH_NAME       : $DASH_NAME
+    DASH_URL        : $DASH_URL
+    DASH_USERNAME   : $DASH_USERNAME
+    DASH_EXTRA_CURL : ${DASH_EXTRA_CURL[@]}
+    VERBOSE         : $VERBOSE
+EOF
+fi
+
+# ================================================================
+# Other prerequisite checks.
+# ================================================================
+if [ ! -f "$DASH_JSON" ] ; then
+    printf "$ERRFMT" $LINENO "file does not exist: $DASH_JSON" 1>&2
+    exit 1
+fi
+if ! grep '"__inputs"' "$DASH_JSON" 1>/dev/null 2>/dev/null ; then
+    printf "$ERRFMT" $LINENO "missing required record '__inputs' in $DASH_JSON" 1>&2
+    exit 1
+fi
+if ! curl --version --version 1>/dev/null 2>/dev/null ; then
+    printf "$ERRFMT" $LINENO "curl is not installed, cannot continue" 1>&2
+    exit 1
+fi
+
+# ================================================================
+# Create the JSON with the variables set.
+#
+# Note the the "inputs" record MUST appear before the "dashboard"
+# record because the import operation picks up the first input
+# specification and ignores the reas.
+#
+# This is relevant because the inputs described below references the
+# correct the datasource where are the __inputs in the dashboard
+# record reference the old, incorrect datasource.
```

Comment:
I'm really not sure what this is trying to say.

Reply:
Yup i see that, i rewrote it thusly:
```python
# This is relevant because there are two sources of external variable
# definitions in the dashboard description. The first is the "inputs"
# record at the top level (shown below) and the other is the
# "__inputs" record under each dashboard (not shown) which is created
# by Grafana when it exports the dashboard. The top level definition
# overrides all subsequent low level definitions which is why it must
# appear first.
```